pkg/ruler/ruler.go

```diff
@@ -166,41 +167,43 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d
 		}
 	}
 
-	workers := make([]worker, cfg.NumWorkers)
 	for i := 0; i < cfg.NumWorkers; i++ {
-		workers[i] = newWorker(ruler)
+		// initialize each worker in a function that signals when
+		// the worker has completed
+		ruler.workers.Add(1)
+		go func() {
+			w := newWorker(ruler)
+			w.Run()
+			ruler.workers.Done()
+		}()
 	}
 
-	ruler.workers = workers
-	go ruler.run()
+	go ruler.scheduler.Run()
 
-	return ruler, nil
-}
-
-// Run the ruler.
-func (r *Ruler) run() {
-	go r.scheduler.Run()
-	for _, w := range r.workers {
-		go w.Run()
-	}
 	level.Info(util.Logger).Log("msg", "ruler up and running")
+
+	return ruler, nil
 }
 
 // Stop stops the Ruler.
+// Each function of the ruler is terminated before leaving the ring
 func (r *Ruler) Stop() {
 	r.notifiersMtx.Lock()
 	for _, n := range r.notifiers {
 		n.stop()
 	}
 	r.notifiersMtx.Unlock()
 
+	level.Info(util.Logger).Log("msg", "ruler shutting down scheduler")
```

Comment:
Do we need the `ruler` here? The logs are coming from a ruler right?

Reply:
It might be nice for it to say `ruler` for single binary mode. The log line should say `ruler.go`, but that has been broken in the past.